phpunit/GutenbergTestSuiteLoader.php

```diff
@@ -0,0 +1,119 @@
+<?php declare(strict_types=1);
+/*
+ * This file is part of PHPUnit.
+ *
+ * (c) Sebastian Bergmann <sebastian@phpunit.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+namespace phpunit;
+
+use PHPUnit\Framework\TestCase;
+use PHPUnit\Util\FileLoader;
+use PHPUnit\Runner\TestSuiteLoader;
+
+use \Exception;
+use \ReflectionClass;
+use \ReflectionException;
+
+/**
+ * @internal This class is not covered by the backward compatibility promise for PHPUnit
+ *
+ * @deprecated see https://github.com/sebastianbergmann/phpunit/issues/4039
+ */
+class GutenbergTestSuiteLoader implements TestSuiteLoader
+{
+	/**
+	 * @throws Exception
+	 */
+	public function load(string $suiteClassFile): ReflectionClass
+	{
+		$suiteClassName = basename($suiteClassFile, '.php');
+		$loadedClasses  = get_declared_classes();
+
+		if (!class_exists($suiteClassName, false)) {
+			/* @noinspection UnusedFunctionResultInspection */
+			FileLoader::checkAndLoad($suiteClassFile);
+
+			$loadedClasses = array_values(
+				array_diff(get_declared_classes(), $loadedClasses)
+			);
+
+			if (empty($loadedClasses)) {
+				throw $this->exceptionFor($suiteClassName, $suiteClassFile);
+			}
+		}
+
+		if (!class_exists($suiteClassName, false)) {
+			$offset = 0 - strlen($suiteClassName);
+
+			foreach ($loadedClasses as $loadedClass) {
+				// @see https://github.com/sebastianbergmann/phpunit/issues/5020
+				if (stripos(substr($loadedClass, $offset - 1), '\\' . $suiteClassName) === 0 ||
+				    stripos(substr($loadedClass, $offset - 1), '_' . $suiteClassName) === 0 ||
+				    strtolower(str_replace('-', '_', $suiteClassName)) === strtolower($loadedClass)) {
+					$suiteClassName = $loadedClass;
+
+					break;
+				}
+			}
+		}
```

Comment:
this entire block confuses me, starting with `$offset = 0 - strlen(…)`

	- why not just `$offset = -strlen(…)`? is `$offset` necessary at all?
	- why are we calling `stripos()` with a `$needle` that has no case variations?
	- why is all this logic implicit when we are building a compound expression?

it looks like we're trying to (a) chop off any namespace prefix, (b) ignore any underlined prefix (so for example `Never_Use_MyClass_Test` would succeed for `MyClass_Test`) and (c) account for variations of case-folding and dash-vs-underscore.

I would imagine this is clearer if we give names to the stages of transformation

```php
function comparableClassName( $className ) {
	// Strip away any leading namespaces: "\My\Custom\Class" -> "Class"
	$className = array_reverse( preg_split( "~\\~", $className ) )[0];
	// Normalize dashes to underscores: "Nifty-Class" -> "Nifty_Class"
	$className = str_replace( '-', '_', $className );
	// Case-fold for case-insensitive comparison
	$className = strtolower( $className );

	return $className;
}

$comparableSuiteClassName = comparableClassName( $suiteClassName );
foreach ( $loadedClasses as $className ) {
	if ( comparableClassName( $className ) === $comparableSuiteClassName ) {
		$suiteClassName = $className;
		break;
	}
}
```

Reply:
All good notes! This update is now included in this PR.